Hopefully/Procedure.cs

```diff
@@ -97,15 +97,22 @@ public static T Retry<T>(Func<T> what, out int failedAttempts, int attempts = 5)
                 {
                     return what();
                 }
-                catch
+                catch(Exception ex)
                 {
-                    attempts--;
-                    failedAttempts++;
-                    if (attempts == 0)
-                        throw;
+                    attempts = FailedAttempt(ref failedAttempts, attempts, wait, ex);   
                 }
             }
             return default(T); // Never reached
         }
+
+        private static int FailedAttempt(ref int failedAttempts, int attempts, TimeSpan wait, Exception ex)
+        {
+            attempts--;
+            failedAttempts++;
+            if (attempts == 0)
+                throw new ApplicationException("Reached the max number of attempts", ex);
+            System.Threading.Thread.Sleep(wait);
```

Comment:
`Thread.Sleep(0)` is not without side effects or performance costs - see [this](https://stackoverflow.com/questions/3257708/thread-sleep0-what-is-the-normal-behavior). Can you wrap this in an `if` statement to avoid calling it if `wait == 0`?